Automattic/jetpack
# CRM: 35359 and 3430 - Add HPOS support to WooSync

Resolves #35359 and Automattic/zero-bs-crm#3430 - add HPOS support to WooSync

## Proposed changes:

This PR:

* Introduces `jpcrm_woosync_is_hpos_enabled()` to check for Woo HPOS. Originally I did it within the WooSync module class, but there are instances where we need it when the class isn't loaded.
* If HPOS, uses `woocommerce_shop_order_list_table_columns` filter instead of `manage_edit-shop_order_columns`.
* If HPOS, uses `woocommerce_shop_order_list_table_custom_column` action instead of `manage_shop_order_posts_custom_column`.
* If HPOS, uses `woocommerce_before_trash_order` action instead of `wp_trash_post`.
* If HPOS, uses `woocommerce_before_delete_order` action instead of `before_delete_post`.
* Updates the three callback functions to account for different param types depending on if HPOS or not.
* Adds a comment to `append_woo_products_to_crm_product_index`, which is not enabled but has non-HPOS calls.

### Other information:


## Jetpack product discussion



## Does this pull request change what data or activity we track or use?



## Testing instructions:





Test on JPCRM + Woo (non-HPOS) as well as JPCRM + Woo (HPOS) to ensure no regressions. HPOS is enabled on all new empty Woo sites; one can disable it per these instructions:

https://woo.com/document/high-performance-order-storage/#section-7

1. Enable WooSync invoice creation: `/wp-admin/admin.php?page=zerobscrm-plugin-settings&tab=woosync`
2. Create a Woo product: `/wp-admin/post-new.php?post_type=product`
3. Create a Woo order with a billing name and email: `/wp-admin/admin.php?page=wc-orders&action=new`
4. Visit the Woo order listview: `/wp-admin/edit.php?post_type=shop_order`
5. Visit the individual Woo order.

In `trunk`, the order listview will only show the CRM contact button on the non-HPOS site, and the order page will only show the contact metabox on the non-HPOS site.

In the `fix/crm/35359-order_metabox_missing_on_HPOS` branch, the contact button and metabox show on the HPOS site as well.

![image](https://github.com/Automattic/jetpack/assets/32492176/cd896812-f973-44e6-928d-06fd86ece7c8)
![image](https://github.com/Automattic/jetpack/assets/32492176/a7d4fc01-447c-443c-a863-bd3b432d5264)
